### ISSUE TITLE
[dask] Cannot use a mix of Array and DataFrame types in training

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#3909]
## Summary

Using the estimators in `lightgbm.dask`, it is not possible to train on a mix of Dask Array and Dask DataFrame. I found, for example, that if `X` is a Dask DataFrame and `y` is a Dask Array, training fails with an error like this

> TypeError: Wrong type(ndarray) for label.
It should be list, numpy 1-D array or pandas Series

<details><summary>full logs</summary>

```text
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-28-e17ddcb3ecf3> in <module>
----> 1 reg.fit(X=ddf, y=label)

~/miniconda3/lib/python3.7/site-packages/lightgbm/dask.py in fit(self, X, y, sample_weight, client, **kwargs)
    720             y=y,
    721             sample_weight=sample_weight,
--> 722             **kwargs
    723         )
    724 

~/miniconda3/lib/python3.7/site-packages/lightgbm/dask.py in _fit(self, model_factory, X, y, sample_weight, group, **kwargs)
    498             sample_weight=sample_weight,
    499             group=group,
--> 500             **kwargs
    501         )
    502 

~/miniconda3/lib/python3.7/site-packages/lightgbm/dask.py in _train(client, data, label, params, model_factory, sample_weight, group, **kwargs)
    337     ]
    338 
--> 339     results = client.gather(futures_classifiers)
    340     results = [v for v in results if v]
    341     return results[0]

~/miniconda3/lib/python3.7/site-packages/distributed/client.py in gather(self, futures, errors, direct, asynchronous)
   1990                 direct=direct,
   1991                 local_worker=local_worker,
-> 1992                 asynchronous=asynchronous,
   1993             )
   1994 

~/miniconda3/lib/python3.7/site-packages/distributed/client.py in sync(self, func, asynchronous, callback_timeout, *args, **kwargs)
    831         else:
    832             return sync(
--> 833                 self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
    834             )
    835 

~/miniconda3/lib/python3.7/site-packages/distributed/utils.py in sync(loop, func, callback_timeout, *args, **kwargs)
    338     if error[0]:
    339         typ, exc, tb = error[0]
--> 340         raise exc.with_traceback(tb)
    341     else:
    342         return result[0]

~/miniconda3/lib/python3.7/site-packages/distributed/utils.py in f()
    322             if callback_timeout is not None:
    323                 future = asyncio.wait_for(future, callback_timeout)
--> 324             result[0] = yield future
    325         except Exception as exc:
    326             error[0] = sys.exc_info()

~/miniconda3/lib/python3.7/site-packages/tornado-6.0.4-py3.7-linux-x86_64.egg/tornado/gen.py in run(self)
    733 
    734                     try:
--> 735                         value = future.result()
    736                     except Exception:
    737                         exc_info = sys.exc_info()

~/miniconda3/lib/python3.7/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1849                             exc = CancelledError(key)
   1850                         else:
-> 1851                             raise exception.with_traceback(traceback)
   1852                         raise exc
   1853                     if errors == "skip":

~/miniconda3/lib/python3.7/site-packages/lightgbm/dask.py in _train_part()
    182             model.fit(data, label, sample_weight=weight, group=group, **kwargs)
    183         else:
--> 184             model.fit(data, label, sample_weight=weight, **kwargs)
    185 
    186     finally:

~/miniconda3/lib/python3.7/site-packages/lightgbm/sklearn.py in fit()
    789                     eval_init_score=eval_init_score, eval_metric=eval_metric,
    790                     early_stopping_rounds=early_stopping_rounds, verbose=verbose, feature_name=feature_name,
--> 791                     categorical_feature=categorical_feature, callbacks=callbacks, init_model=init_model)
    792         return self
    793 

~/miniconda3/lib/python3.7/site-packages/lightgbm/sklearn.py in fit()
    632                               evals_result=evals_result, fobj=self._fobj, feval=eval_metrics_callable,
    633                               verbose_eval=verbose, feature_name=feature_name,
--> 634                               callbacks=callbacks, init_model=init_model)
    635 
    636         if evals_result:

~/miniconda3/lib/python3.7/site-packages/lightgbm/engine.py in train()
    226     # construct booster
    227     try:
--> 228         booster = Booster(params=params, train_set=train_set)
    229         if is_valid_contain_train:
    230             booster.set_train_data_name(train_data_name)

~/miniconda3/lib/python3.7/site-packages/lightgbm/basic.py in __init__()
   2211                 )
   2212             # construct booster object
-> 2213             train_set.construct()
   2214             # copy the parameters from train_set
   2215             params.update(train_set.get_params())

~/miniconda3/lib/python3.7/site-packages/lightgbm/basic.py in construct()
   1454                                 init_score=self.init_score, predictor=self._predictor,
   1455                                 silent=self.silent, feature_name=self.feature_name,
-> 1456                                 categorical_feature=self.categorical_feature, params=self.params)
   1457             if self.free_raw_data:
   1458                 self.data = None

~/miniconda3/lib/python3.7/site-packages/lightgbm/basic.py in _lazy_init()
   1264                 raise TypeError('Cannot initialize Dataset from {}'.format(type(data).__name__))
   1265         if label is not None:
-> 1266             self.set_label(label)
   1267         if self.get_label() is None:
   1268             raise ValueError("Label should not be None")

~/miniconda3/lib/python3.7/site-packages/lightgbm/basic.py in set_label()
   1783         self.label = label
   1784         if self.handle is not None:
-> 1785             label = list_to_1d_numpy(_label_from_pandas(label), name='label')
   1786             self.set_field('label', label)
   1787             self.label = self.get_field('label')  # original values can be modified at cpp side

~/miniconda3/lib/python3.7/site-packages/lightgbm/basic.py in list_to_1d_numpy()
    147     else:
    148         raise TypeError("Wrong type({0}) for {1}.\n"
--> 149                         "It should be list, numpy 1-D array or pandas Series".format(type(data).__name__, name))
    150 
    151 

TypeError: Wrong type(ndarray) for label.
It should be list, numpy 1-D array or pandas Series

```

</details>

## Notes

I haven't investigated this at all yet. Just noticed it while testing #3908 and wanted to write it up.

## Reproducible Examples

This can be reproduced with the example below.

```python
import dask
import dask.array as da
import dask.dataframe as dd
import lightgbm as lgb
import numpy as np
import pandas as pd

from distributed import Client, LocalCluster

cluster = LocalCluster(n_workers=3)
client = Client(cluster)


def _create_data() -> pd.DataFrame:
    num_rows = 1000
    return pd.DataFrame({
        "float_col1": pd.Series(np.random.random(num_rows), dtype="float"),
        "float_col2": pd.Series(np.random.random(num_rows), dtype="float"),
        "cat_col": pd.Series(np.random.choice(["a", "b", "y", "z"], num_rows), dtype="category"),
    })


parts = [dask.delayed(_create_data)() for _ in range(5)]
ddf = dd.from_delayed(
    parts,
    meta={
        "float_col1": "float",
        "float_col2": "float",
        "cat_col": "category"
    }
)

label = da.random.random((5000, 1), (1000, 1))
reg = lgb.DaskLGBMRegressor()
reg.fit(X=ddf, y=label)
```

## How you are using LightGBM?

LightGBM component: Python Package

## Environment info

Operating System: Ubuntu 18.04

C++ compiler version: `gcc` 7.5.0

CMake version: 3.10.2

Python version: see `conda info` output below

<details><summary>conda info (click me)</summary>

```text
     active environment : None
       user config file : /home/jlamb/.condarc
 populated config files : 
          conda version : 4.9.2
    conda-build version : not installed
         python version : 3.7.6.final.0
       virtual packages : __glibc=2.27=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/jlamb/miniconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/jlamb/miniconda3/pkgs
                          /home/jlamb/.conda/pkgs
       envs directories : /home/jlamb/miniconda3/envs
                          /home/jlamb/.conda/envs
               platform : linux-64
             user-agent : conda/4.9.2 requests/2.22.0 CPython/3.7.6 Linux/5.4.0-65-generic ubuntu/18.04.4 glibc/2.27
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

</details>

LightGBM version or commit hash: latest on `master`, https://github.com/microsoft/LightGBM/tree/fc2da2ec84e2dd9aa06421edcb5a625808eb546e

**Comment by jmoralez** [comment]
I believe this is actually because `label` has shape (5000,1) instead of just (5000,). I just tried it with 
`label = da.random.random(5000, chunks=1000)` and it works.

**Comment by jameslamb** [comment]
Thanks for investigating @jmoralez ! But I don't think that's exactly the problem.

If you try changing `ddf` to a Dask array and leave `label` exactly as it is specified in my example, you'll get a warning

> /opt/conda/lib/python3.8/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
/opt/conda/lib/python3.8/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)

but training will still succeed

```python
import dask
import dask.array as da
import dask.dataframe as dd
import lightgbm as lgb
import numpy as np
import pandas as pd

from distributed import Client, LocalCluster

cluster = LocalCluster(n_workers=3)
client = Client(cluster)


def _create_data() -> pd.DataFrame:
    num_rows = 1000
    return pd.DataFrame({
        "float_col1": pd.Series(np.random.random(num_rows), dtype="float"),
        "float_col2": pd.Series(np.random.random(num_rows), dtype="float"),
    })

parts = [dask.delayed(_create_data)() for _ in range(5)]
ddf = dd.from_delayed(
    parts,
    meta={
        "float_col1": "float",
        "float_col2": "float",
    }
)

label = da.random.random((5000, 1), (1000, 1))
reg = lgb.DaskLGBMRegressor()
reg.fit(X=ddf.to_dask_array(), y=label)
```

**Comment by jmoralez** [comment]
I've investigated a bit more and the type of `X` decides which type of validation to perform on the inputs [here](https://github.com/microsoft/LightGBM/blob/fc2da2ec84e2dd9aa06421edcb5a625808eb546e/python-package/lightgbm/sklearn.py#L561). So when `X` is an array it uses scikit-learns [check_X_y](https://github.com/scikit-learn/scikit-learn/blob/819c43cc7a1d7efab855e982a91e15be7aec7db1/sklearn/utils/validation.py#L708) which can detect an (n_samples, 1) array and flatten it. However, if `X` is a dataframe the validation is skipped and it ends up trying to parse the label [here](https://github.com/microsoft/LightGBM/blob/fc2da2ec84e2dd9aa06421edcb5a625808eb546e/python-package/lightgbm/basic.py#L1785) when calling `Dataset.construct`. I guess this could be solved by adding a case for when an array like this is sent as label to the [list_to_1d_numpy](https://github.com/microsoft/LightGBM/blob/fc2da2ec84e2dd9aa06421edcb5a625808eb546e/python-package/lightgbm/basic.py#L134) function.

**Comment by jmoralez** [comment]
Should I go for it?

**Comment by jameslamb** [comment]
sure!

**Comment by jmoralez** [comment]
Hi James. I've added the case for the column array and it works, however I'm not sure how to assert the warning since it happens in the workers. Do you have any suggestions for this?
I can see this in ipython:
```
/home/josemz/programs/anaconda3/envs/lightgbm/lib/python3.7/site-packages/lightgbm/basic.py:153: UserWarning: Converting column vector to 1d array
  _log_warning('Converting column vector to 1d array')                                                                                            
/home/josemz/programs/anaconda3/envs/lightgbm/lib/python3.7/site-packages/lightgbm/basic.py:153: UserWarning: Converting column vector to 1d array
  _log_warning('Converting column vector to 1d array')                                                                                            
/home/josemz/programs/anaconda3/envs/lightgbm/lib/python3.7/site-packages/lightgbm/basic.py:153: UserWarning: Converting column vector to 1d array
  _log_warning('Converting column vector to 1d array')                                                                                            
/home/josemz/programs/anaconda3/envs/lightgbm/lib/python3.7/site-packages/lightgbm/basic.py:153: UserWarning: Converting column vector to 1d array
  _log_warning('Converting column vector to 1d array')
```
However I can't assert the `pytest.warns`

**Comment by jmoralez** [comment]
Actually now that I think about it this isn't a dask specific problem, this happens with standard LightGBM as well, should I move the test there?

**Comment by jameslamb** [comment]
> this isn't a dask specific problem, this happens with standard LightGBM as well, should I move the test there?

If you think the underlying problem is not in the Dask module that's fine, but any pull request submitted to fix it should also include a unit test in the Dask tests, please. To be sure that it is fixed there and that we don't accidentally re-introduce the issue in the future.

**Comment by jameslamb** [comment]
fixed by #3943
